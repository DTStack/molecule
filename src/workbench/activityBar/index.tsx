import 'reflect-metadata';
import { connect } from 'mo/react';
import { ActivityBarController } from 'mo/controller';

import ActivityBar from './activityBar';
import { container } from 'tsyringe';
import { ActivityBarService } from 'mo/services';
export * from './activityBar';
export { ActivityBarItem } from './activityBarItem';

const activityBarService = container.resolve(ActivityBarService);
const activityBarController = container.resolve(ActivityBarController);

export const ActivityBarView = connect(
    activityBarService,
    ActivityBar,
    activityBarController
);
