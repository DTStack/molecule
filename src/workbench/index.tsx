export { WorkbenchView as Workbench } from './workbench';
export { ActivityBar } from 'mo/workbench/activityBar';
export { ActivityBarItem } from 'mo/workbench/activityBar/activityBarItem';
export { Editor } from 'mo/workbench/editor/editor';
export { EditorGroup } from 'mo/workbench/editor/group';
export { MenuBar } from 'mo/workbench/menuBar/menuBar';
export { Panel } from 'mo/workbench/panel/panel';
export { Sidebar } from 'mo/workbench/sidebar';
export { StatusBar, StatusItem } from 'mo/workbench/statusBar';

export type {
    IWorkbench,
    IActivityBar,
    IActivityBarItem,
    IEditor,
    IEditorGroup,
    IMenuBar,
    IPanel,
    ISidebar,
    IStatusBar,
    IStatusBarItem,
    IProblems,
    IProblemsItem,
    INotification,
    INotificationItem,
} from 'mo/model';

export type {
    IActivityBarController,
    ILayoutController,
    IEditorController,
    ISideBarController,
    IMenuBarController,
    ISettingsController,
    INotificationController,
    IPanelController,
    IProblemsController,
    IStatusBarController,
} from 'mo/controller';
