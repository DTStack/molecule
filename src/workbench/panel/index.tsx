import 'reflect-metadata';
import { container } from 'tsyringe';
import { connect } from 'mo/react';
import { IPanelService, PanelService } from 'mo/services';
import Panel from './panel';
import { PanelController } from 'mo/controller/panel';

const panelService = container.resolve<IPanelService>(PanelService);
const panelController = container.resolve(PanelController);

const PanelView = connect(panelService, Panel, panelController);

export { PanelView };
