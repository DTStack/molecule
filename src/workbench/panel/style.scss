@import 'mo/style/common';
$icon-size: 35px;

#{$panel} {
    background-color: var(--panel-background);
    border-color: var(--panel-border);
    display: flex;
    height: 100%;
    position: relative;
    width: 100%;

    &__header,
    &__tabs,
    &__toolbar {
        display: flex;
    }

    &__header {
        border-color: inherit;
        border-top-style: solid;
        border-top-width: 1px;
        flex-direction: row;
        height: 35px;
        justify-content: space-between;
        padding: 0 8px;
        width: 100%;

        #{$tabs} {
            background: inherit;
            border: 0;
            flex-grow: 1;
        }

        #{$tab}__item {
            background: inherit;
            border: 0;
            color: var(--panelTitle-inactiveForeground);
            font-size: 11px;
            padding: 0 10px;
            text-transform: uppercase;

            &--active {
                border-color: var(--panelTitle-activeBorder);
                color: var(--panelTitle-activeForeground);

                &::after {
                    background-color: var(--panelTitle-activeBorder);
                    bottom: 2px;
                    content: ' ';
                    height: 1px;
                    left: 10px;
                    position: absolute;
                    right: 10px;
                }
            }
        }
    }

    &__container {
        border: 1px solid transparent;
        bottom: 0;
        box-sizing: border-box;
        left: 0;
        position: absolute;
        right: 0;
        top: 35px;
        width: 100%;

        &:focus {
            border-color: var(--list-focusOutline);
        }
    }

    &__toolbar {
        .codicon {
            height: $icon-size;
            line-height: $icon-size;
            margin-right: 4px;
            min-width: 28px;
        }
    }
}

#{$output} {
    height: 100%;
    width: 100%;
}
