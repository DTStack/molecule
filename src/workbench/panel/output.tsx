import React from 'react';
import { prefixClaName } from 'mo/common/className';
import { IOutput } from 'mo/model/workbench/panel';
import { MonacoEditor } from 'mo/components/monaco';

const defaultClassName = prefixClaName('output');

function Output(props: IOutput) {
    const { id, data = '', onUpdateEditorIns } = props;

    return (
        <div className={defaultClassName}>
            <MonacoEditor
                key={id}
                options={{
                    value: data,
                    readOnly: true,
                    lineDecorationsWidth: 0,
                    lineNumbers: 'off',
                    minimap: {
                        enabled: false,
                    },
                    automaticLayout: true,
                }}
                editorInstanceRef={(editorInstance) => {
                    onUpdateEditorIns?.(editorInstance);
                }}
            />
        </div>
    );
}

export default Output;
