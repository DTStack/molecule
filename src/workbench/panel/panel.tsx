import React from 'react';
import { getBEMElement, prefixClaName } from 'mo/common/className';
import { IPanel, IPanelItem } from 'mo/model/workbench/panel';
import { IPanelController } from 'mo/controller/panel';
import { Tabs } from 'mo/components/tabs';
import { ActionBar } from 'mo/components/actionBar';
import { Scrollable } from 'mo/components';

const defaultClassName = prefixClaName('panel');
const panelHeaderClassName = getBEMElement(defaultClassName, 'header');
const panelToolbarClassName = getBEMElement(defaultClassName, 'toolbar');
const panelContainerClassName = getBEMElement(defaultClassName, 'container');

export function Panel(props: IPanel & IPanelController) {
    const { data, current, toolbox, onTabChange, onToolbarClick, onClose } =
        props;
    let toolboxData = toolbox || [];
    if (current && current.toolbox) {
        toolboxData = current.toolbox.concat(toolboxData);
    }

    const content =
        typeof current?.renderPane === 'function'
            ? current?.renderPane?.(current)
            : current?.renderPane;

    const sortedPanels = data?.sort((a: IPanelItem, b: IPanelItem) => {
        if (a.sortIndex && b.sortIndex) {
            return a.sortIndex - b.sortIndex;
        }
        return 0;
    });

    return (
        <div className={defaultClassName}>
            <div className={panelHeaderClassName}>
                <Scrollable noScrollY>
                    <Tabs
                        activeTab={current?.id}
                        data={sortedPanels}
                        onSelectTab={onTabChange}
                        onCloseTab={onClose}
                    />
                </Scrollable>
                <ActionBar
                    className={panelToolbarClassName}
                    data={toolboxData}
                    onClick={onToolbarClick}
                />
            </div>
            <div className={panelContainerClassName} tabIndex={0}>
                {content}
            </div>
        </div>
    );
}

export default Panel;
