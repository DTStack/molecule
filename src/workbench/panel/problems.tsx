import * as React from 'react';
import { memo } from 'react';
import { prefixClaName } from 'mo/common/className';
import { IPanelItem } from 'mo/model/workbench/panel';

const defaultClassName = prefixClaName('problems');

function Problems(props: IPanelItem) {
    const { data } = props;
    return (
        <div className={defaultClassName} style={{ margin: '0 18px' }}>
            Problems {data}
        </div>
    );
}

export default memo(Problems);
