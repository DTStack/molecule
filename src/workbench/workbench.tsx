import 'reflect-metadata';
import React from 'react';
import { container } from 'tsyringe';

import {
    classNames,
    getFontInMac,
    prefixClaName,
    getBEMModifier,
    getBEMElement,
} from 'mo/common/className';

import { EditorView } from 'mo/workbench/editor';
import { SidebarView } from 'mo/workbench/sidebar';
import { MenuBarView } from 'mo/workbench/menuBar';
import { ActivityBarView } from 'mo/workbench/activityBar';
import { StatusBarView } from 'mo/workbench/statusBar';
import { PanelView } from 'mo/workbench/panel';
import { ID_APP } from 'mo/common/id';
import { APP_PREFIX } from 'mo/common/const';

import { connect } from 'mo/react';

import { ILayoutController, LayoutController } from 'mo/controller/layout';
import { LayoutService } from 'mo/services';
import { ILayout, MenuBarMode } from 'mo/model/workbench/layout';

import { IWorkbench } from 'mo/model';
import SplitPane from 'mo/components/split/SplitPane';
import { Pane } from 'mo/components/split';
import { Display } from 'mo/components';

const mainBenchClassName = prefixClaName('mainBench');
const workbenchClassName = prefixClaName('workbench');
const compositeBarClassName = prefixClaName('compositeBar');
const appClassName = classNames(APP_PREFIX, getFontInMac());
const workbenchWithHorizontalMenuBarClassName = getBEMModifier(
    workbenchClassName,
    'with-horizontal-menuBar'
);
const withHiddenStatusBar = getBEMModifier(
    workbenchClassName,
    'with-hidden-statusBar'
);
const displayActivityBarClassName = getBEMElement(
    workbenchClassName,
    'display-activityBar'
);

const layoutController = container.resolve(LayoutController);
const layoutService = container.resolve(LayoutService);

export function WorkbenchView(props: IWorkbench & ILayout & ILayoutController) {
    const {
        activityBar,
        menuBar,
        panel,
        sidebar,
        statusBar,
        onPaneSizeChange,
        onHorizontalPaneSizeChange,
        splitPanePos,
        horizontalSplitPanePos,
    } = props;

    const getSizes = () => {
        if (panel.hidden) {
            return ['100%', 0];
        }
        if (panel.panelMaximized) {
            return [0, '100%'];
        }
        return horizontalSplitPanePos;
    };

    const isMenuBarVertical =
        !menuBar.hidden && menuBar.mode === MenuBarMode.vertical;
    const isMenuBarHorizontal =
        !menuBar.hidden && menuBar.mode === MenuBarMode.horizontal;
    const horizontalMenuBar = isMenuBarHorizontal
        ? workbenchWithHorizontalMenuBarClassName
        : null;
    const hideStatusBar = statusBar.hidden ? withHiddenStatusBar : null;
    const workbenchFinalClassName = classNames(
        workbenchClassName,
        horizontalMenuBar,
        hideStatusBar
    );

    return (
        <div id={ID_APP} className={appClassName} tabIndex={0}>
            <div className={workbenchFinalClassName}>
                <Display visible={isMenuBarHorizontal}>
                    <MenuBarView mode={MenuBarMode.horizontal} />
                </Display>
                <div className={mainBenchClassName}>
                    <div className={compositeBarClassName}>
                        <Display visible={isMenuBarVertical}>
                            <MenuBarView mode={MenuBarMode.vertical} />
                        </Display>
                        <Display
                            visible={!activityBar.hidden}
                            className={displayActivityBarClassName}
                        >
                            <ActivityBarView />
                        </Display>
                    </div>
                    <SplitPane
                        sizes={sidebar.hidden ? [0, '100%'] : splitPanePos}
                        split="vertical"
                        showSashes={!sidebar.hidden}
                        allowResize={[false]}
                        onChange={onPaneSizeChange!}
                    >
                        <Pane minSize={170} maxSize="80%">
                            <SidebarView />
                        </Pane>
                        <SplitPane
                            sizes={getSizes()}
                            showSashes={!panel.hidden && !panel.panelMaximized}
                            allowResize={[true, false]}
                            split="horizontal"
                            onChange={onHorizontalPaneSizeChange!}
                        >
                            <Pane minSize="10%" maxSize="80%">
                                <EditorView />
                            </Pane>
                            <PanelView />
                        </SplitPane>
                    </SplitPane>
                </div>
            </div>
            <Display visible={!statusBar.hidden}>
                <StatusBarView />
            </Display>
        </div>
    );
}

export const Workbench = connect(
    layoutService,
    WorkbenchView,
    layoutController
);
