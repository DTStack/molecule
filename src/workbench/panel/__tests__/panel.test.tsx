import 'reflect-metadata';
import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';

import { PanelView } from '../index';
import { Panel } from '../panel';
import { IPanel, PanelModel } from 'mo/model/workbench/panel';
import { select } from 'mo/common/dom';
import { modules } from 'mo/services/builtinService/const';
import { cloneDeep } from 'lodash';
import Output from '../output';
import { container } from 'tsyringe';
import { ExtendsLocales } from 'mo/extensions/locales-defaults';
import { LocaleService } from 'mo/i18n';

function panelMockModel(): PanelModel {
    const output = modules.builtInOutputPanel();
    output.renderPane = (item) => (
        <Output
            onUpdateEditorIns={(instance) => {
                // Please notice the problem about memory out
                // 'Cause we didn't dispose the older instance
                item.outputEditorInstance = instance;
            }}
            {...item}
        />
    );
    const problems = modules.builtInPanelProblems();
    const toolboxResize = [
        modules.builtInPanelToolboxResize(),
        modules.builtInPanelToolbox(),
    ];
    return new PanelModel(output, [output, problems], toolboxResize);
}

describe('Test Panel Component', () => {
    // initial locales
    const localeService = container.resolve(LocaleService);
    localeService.initialize(ExtendsLocales.contributes!.languages!, 'en');

    test('Match the PanelView snapshot', () => {
        const component = renderer.create(<PanelView />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    test('Test PanelView render', () => {
        render(<PanelView />);
        expect(select<HTMLDivElement>('.mo-tab__item')).toBeNull();
    });

    test('Match the Panel snapshot', () => {
        const component = renderer.create(<Panel {...panelMockModel()} />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    test('Customize the Panel toolbox', () => {
        const panel: IPanel = new PanelModel();
        const output = cloneDeep(modules.builtInOutputPanel());
        output.renderPane = (item) => (
            <Output
                onUpdateEditorIns={(instance) => {
                    // Please notice the problem about memory out
                    // 'Cause we didn't dispose the older instance
                    item.outputEditorInstance = instance;
                }}
                {...item}
            />
        );
        panel.toolbox = [
            modules.builtInPanelToolboxResize(),
            modules.builtInPanelToolbox(),
        ];
        panel.current = output;
        const { queryByText, rerender, queryByTestId } = render(
            <Panel {...panel} />
        );
        expect(queryByText('box1')).not.toBeInTheDocument();

        output.toolbox = [{ id: 'box1', name: 'box1', 'data-testid': 'box1' }];
        rerender(<Panel {...panel} />);
        expect(queryByTestId('box1')).toBeInTheDocument();

        document.body.innerHTML = '';
        panel.toolbox = undefined;
        rerender(<Panel {...panel} />);
        expect(queryByTestId('box1')).not.toBeInTheDocument();
    });

    test('Customize the Panel render content', () => {
        const panel = panelMockModel();
        const { container, rerender, getAllByText } = render(
            <Panel {...panel} />
        );
        expect(
            container.querySelector('.mo-monaco-editor')
        ).toBeInTheDocument();

        panel.current!.renderPane = () => <span>customizedPane</span>;
        rerender(<Panel {...panel} />);
        expect(
            container.querySelector('.mo-monaco-editor')
        ).not.toBeInTheDocument();
        expect(getAllByText('customizedPane').length).toBe(1);
    });

    test('Sort the Panel ', () => {
        const current = { id: '1', name: 'test1' };
        const panel2 = { id: '2', name: 'test2' };

        const panel = new PanelModel(current, [current, panel2]);

        const { container, rerender } = render(<Panel {...panel} />);
        const tabs =
            container.querySelectorAll<HTMLDivElement>('.mo-tab__item');
        expect(tabs![0].textContent).toEqual('test1');
        expect(tabs![1].textContent).toEqual('test2');

        panel.data[0].sortIndex = 2;
        panel.data[1].sortIndex = 1;

        rerender(<Panel {...panel} />);
        const tabs1 = container.querySelector<HTMLDivElement>('.mo-tab__item');
        expect(tabs1!.textContent).toEqual('test2');
    });
});
