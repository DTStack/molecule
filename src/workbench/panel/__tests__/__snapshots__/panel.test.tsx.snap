// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Panel Component Match the Panel snapshot 1`] = `
<div
  className="mo-panel"
>
  <div
    className="mo-panel__header"
  >
    <div
      className="mo-tabs mo-tabs--line"
    >
      <div
        className="mo-tabs__header"
      >
        <div
          className="mo-tab__item"
          onClick={[Function]}
          onContextMenu={[Function]}
        >
          Problems
        </div>
        <div
          className="mo-tab__item mo-tab__item--active"
          onClick={[Function]}
          onContextMenu={[Function]}
        >
          output
        </div>
      </div>
    </div>
    <div
      className="mo-action-bar mo-panel__toolbar"
    >
      <ul
        className="mo-action-bar__container"
      >
        <li
          className="mo-action-bar__item"
          data-index={0}
          onClick={[Function]}
        >
          <span
            className="codicon codicon-chevron-up"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          />
        </li>
        <li
          className="mo-action-bar__item"
          data-index={1}
          onClick={[Function]}
        >
          <span
            className="codicon codicon-close"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          />
        </li>
      </ul>
    </div>
  </div>
  <div
    className="mo-panel__container"
    tabIndex={0}
  >
    <div
      className="mo-output"
    >
      <div
        className="mo-monaco-editor"
        style={
          Object {
            "height": "100%",
            "minHeight": "400px",
            "position": "relative",
            "width": "100%",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`Test Panel Component Match the PanelView snapshot 1`] = `
<div
  className="mo-panel"
>
  <div
    className="mo-panel__header"
  >
    <div
      className="mo-tabs mo-tabs--line"
    >
      <div
        className="mo-tabs__header"
      />
    </div>
    <div
      className="mo-action-bar mo-panel__toolbar"
    >
      <ul
        className="mo-action-bar__container"
      />
    </div>
  </div>
  <div
    className="mo-panel__container"
    tabIndex={0}
  />
</div>
`;
